htdp-lib/2htdp/abstraction.rkt

```diff
@@ -114,7 +115,33 @@
     [(_ type:id case:expr [variant:id (field:id ...) body:expr] ...)
      #'(type-case type case [variant (field ...) body] ...)]))
 
+(begin-for-syntax
+  (define-syntax-class pattern-cls
+    #:description "pattern"
+    #:attributes (expr)
+    (pattern name:id
+             #:with expr #'name)
+
+    ;; literal-constants
+    (pattern {~or* :string :number :boolean :char :regexp :bytes :byte-regexp}
```

Comment:
Do you really want to allow the weirder kinds of constants here (eg, byte-regexps)?